text/3334-niche.md

```diff
@@ -0,0 +1,482 @@
+- Feature Name: `niche`
+- Start Date: 2022-10-16
+- RFC PR: [rust-lang/rfcs#3334](https://github.com/rust-lang/rfcs/pull/3334)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Provide a stable attribute to define "niche" values of a type. The type cannot
+store these values, allowing the compiler to use them to optimize the
+representation of containing types such as `Option<Type>`.
+
+# Motivation
+[motivation]: #motivation
+
+Rust makes extensive use of types like `Option`, and many programs benefit from
+the efficient storage of such types. Many programs also interface with others
+via FFI, via interfaces that provide data and a sentinel value (such as for
+errors or missing data) within the same bits.
+
+The Rust compiler already provides support for this via "niche" optimizations,
+and various types providing guarantees of such optimizations, including
+references, `bool`, `char`, and the `NonZero` family of types. However, Rust
+does not provide any stable means of defining new types with niches, reserving
+this mechanism for the standard library. This puts pressure on the standard
+library to provide additional families of types with niches, while preventing
+the broader crate ecosystem from experimenting with such types.
+
+Past efforts to define a stable niche mechanism stalled out due to scope creep:
+alignment niches, null-page niches, multiple niches, structures with multiple
+fields, and many other valid but challenging ideas (documented in the "Future
+possibilities" section). This RFC defines a *simple* mechanism for defining one
+common type of niche, while leaving room for future extension.
+
+Defining a niche mechanism allows libraries to build arbitrary types containing
+niches, and simplifies handling of space-efficient data structures in Rust
```

Comment:
You mentioned in one of the other comment threads that

> Even if we want to provide types that use const generics to represent value ranges, we need some means of creating those types. I would like to give the ecosystem the tools we're using, not just things we've built with those tools. (See the motivation of the RFC.)

Can you elaborate on that more?

To me, the goal of "allow[ing] libraries to build arbitrary types containing niches" is perfectly well satisfied by letting them create newtypes over `core` types with customizable niches.

And in any situation in which the value-with-niche isn't the whole representation of the public type (like if a couple different enum variants have private fields with various niches), the library type is substantially more convenient than needing to `#[niche] #[derive(Copy, Clone, Eq, PartialEq, Hash)]` a couple single-use internal types.  (And this is even more true now that this RFC doesn't even propose having the `repr(transparent)` rules, since I can't even put a `PhantomData` in the type, and thus `#[repr(transparent)] pub struct IndexForArrayOf<T>(pub RefinedUsize<0, {isize::MAX as usize}>, PhantomData<fn(T)>); is more ergonomic than needing both that type *and* my own custom `#[niche]` type.)

Much as I dislike `repr(packed)`, that one at least has the ergonomic justification that [one attribute](https://github.com/retep998/winapi-rs/blob/2f76bdea3a79817ccfab496fbd1786d5a697387b/src/shared/usbspec.rs#L49) is *much* easier than something on every field like if it needed to be
```rust
#[repr(C)]
struct USB_DEFAULT_PIPE_SETUP_PACKET {
    bmRequestType: Packed<BM_REQUEST_TYPE>,
    bRequest: Packed<UCHAR>,
    wValue: Packed<USB_DEFAULT_PIPE_SETUP_PACKET_wValue>,
    wIndex: Packed<USB_DEFAULT_PIPE_SETUP_PACKET_wIndex>,
    wLength: Packed<USHORT>,
}
```

But defining one combined mega-niche for a multi-field struct seems *way* more confusing than putting niches on each field, and if they're on fields then specifying them via const generics on a type seems completely fine.